ngraph/core/include/openvino/core/extension.hpp

```diff
@@ -0,0 +1,73 @@
+// Copyright (C) 2018-2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include <memory>
+#include <type_traits>
+#include <vector>
+
+#include "openvino/core/core_visibility.hpp"
+#include "openvino/core/type.hpp"
+
+// Use extern "C" in order to avoid issues with mangling
+#if defined(_WIN32) && defined(IMPLEMENT_OPENVINO_EXTENSION_API)
+#    define OPENVINO_EXTENSION_C_API extern "C" __declspec(dllexport)
+#    define OPENVINO_EXTENSION_API   __declspec(dllexport)
+#else
+#    define OPENVINO_EXTENSION_C_API extern "C" OPENVINO_API
+#    define OPENVINO_EXTENSION_API   OPENVINO_API
+#endif
+
+namespace ov {
+
+class OPENVINO_API Extension {
```

Comment:
enable shared_from_this

Reply:
Why do we need to have shared_from_this for extensions?

Reply:
It is intended to be used just with `std::shared_ptr` is not it? So it is just trivial optimization to place reference_cointer inside Extension class and reduce heap fragmentation a little bit in cases when it is created not by std::make_shared.